File: backend/app/routers/posts.py
"""Post listing — powers the Facebook drill-down (post -> its comments)."""
from __future__ import annotations

from datetime import datetime

from fastapi import APIRouter, Depends, Query
from sqlalchemy import case, func, select
from sqlalchemy.orm import Session

from ..database import get_db
from ..models import Analysis, Comment, Post

router = APIRouter(prefix="/api/posts", tags=["posts"])


@router.get("")
def list_posts(
    db: Session = Depends(get_db),
    source: str | None = None,
    date_from: datetime | None = Query(None, alias="from"),
    date_to: datetime | None = Query(None, alias="to"),
    limit: int = 100,
):
    stmt = (
        select(
            Post.id, Post.source_type, Post.external_id, Post.published_at,
            Post.message, Post.permalink,
            func.count(Comment.id),
            func.sum(case((Analysis.sentiment == "negative", 1), else_=0)),
        )
        .outerjoin(Comment, Comment.post_id == Post.id)
        .outerjoin(Analysis, Analysis.comment_id == Comment.id)
        .group_by(Post.id)
        .order_by(Post.published_at.desc().nullslast())
        .limit(limit)
    )
    if source:
        stmt = stmt.where(Post.source_type == source)
    if date_from:
        stmt = stmt.where(Post.published_at >= date_from)
    if date_to:
        stmt = stmt.where(Post.published_at <= date_to)

    rows = db.execute(stmt).all()
    return [
        {
            "id": r[0], "source_type": r[1], "external_id": r[2],
            "published_at": r[3].isoformat() if r[3] else None,
            "message": r[4], "permalink": r[5],
            "comment_count": r[6] or 0, "negative_count": int(r[7] or 0),
        }
        for r in rows
    ]

File: worker/src/routes/posts.ts
import { Hono } from "hono";
import type { Env } from "../types";

export const postsRoute = new Hono<{ Bindings: Env }>();

function dateKey(value?: string | null) {
  const text = String(value || "").slice(0, 10);
  return /^\d{4}-\d{2}-\d{2}$/.test(text) ? text : null;
}

function addDateFilter(where: string[], params: any[], column: string, operator: ">=" | "<=", value?: string) {
  const key = dateKey(value);
  if (!key) return;
  where.push(`substr(${column}, 1, 10) ${operator} ?`);
  params.push(key);
}

postsRoute.get("/", async (c) => {
  const q = c.req.query();
  const where: string[] = [];
  const params: any[] = [];
  if (q.source) { where.push("p.source_type = ?"); params.push(q.source); }
  addDateFilter(where, params, "p.published_at", ">=", q.from);
  addDateFilter(where, params, "p.published_at", "<=", q.to);
  const whereSql = where.length ? `WHERE ${where.join(" AND ")}` : "";
  const limit = Math.max(1, Number(q.limit) || 100);

  const rows = await c.env.DB
    .prepare(
      `SELECT p.id, p.source_type, p.external_id, p.published_at, p.message, p.permalink,
              COUNT(cm.id) as comment_count,
              SUM(CASE WHEN a.sentiment = 'negative' THEN 1 ELSE 0 END) as negative_count
       FROM posts p
       LEFT JOIN comments cm ON cm.post_id = p.id
       LEFT JOIN analyses a ON a.comment_id = cm.id
       ${whereSql}
       GROUP BY p.id
       ORDER BY p.published_at DESC
       LIMIT ?`
    )
    .bind(...params, limit)
    .all();

  return c.json(
    rows.results.map((r: any) => ({
      id: r.id,
      source_type: r.source_type,
      external_id: r.external_id,
      published_at: r.published_at,
      message: r.message,
      permalink: r.permalink,
      comment_count: r.comment_count || 0,
      negative_count: Number(r.negative_count || 0),
    }))
  );
});
